### PR TITLE
feat(sink): update sink writer trait to provide richer runtime context

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

In this PR, we do the following change on the sink writer trait so that the sink writer can get richer context
1. We introduce a new `SinkWriter` trait. The new trait will call `start_epoch` for every new epoch, and will be notified on every barrier. Previous trait is renamed to `SinkWriterV1`. The previous trait only `start_epoch` for the first epoch after a checkpoint rather than every epoch, and only get notified on the checkpoint barrier if the checkpoint is non-empty, and will not get notified if the checkpoint is empty. For backward compatibility, we implement a `SinkWriterV1Adapter` so that all previously implemented sink writer can be untouched under the new trait.
2. We introduce a `SinkWriterParam` to provide more runtime information when creating the sink writer. Specifically, previously we have an `identifier` field in kafka config, which is generated by the executor id at runtime. However, as a config, it should be statically specified by users instead of being generated at runtime, and currently we have to manually insert the executor id to the user specified sink properties, which can easily cause error. Therefore, in this PR, we remove the `identifier` from the kafka config field and instead fetch the executor id from the `SinkWriterParam` provided at runtime.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/connector/src/sink/kafka.rs [hunk 1]
```diff
@@ -220,10 +223,12 @@ impl KafkaSinkWriter {
         schema: Schema,
         pk_indices: Vec<usize>,
         is_append_only: bool,
+        identifier: String,
     ) -> Result<Self> {
         Ok(KafkaSinkWriter {
             config: config.clone(),
-            conductor: KafkaTransactionConductor::new(config).await?,
+            conductor: KafkaTransactionConductor::new(config, &identifier).await?,
```
Comment (tabVersion) [inline src/connector/src/sink/kafka.rs:230]: how about removing the abstraction? After decoupling, we can tolerant failure from network jitter, there is no need for a retry layer.

**Review by tabVersion — APPROVED** [review]
the new trait LGTM

File: src/connector/src/sink/kafka.rs (hunk 1 above)
Comment (wenym1) [reply src/connector/src/sink/kafka.rs:230]: LGTM. I think we can do this in another PR.

**Comment by StrikeW** [comment]
I am confused about how "backward compatibility" is defined here. Since our sink writer is not pluggable, each time we upgrade a compute node, all sink writers should be upgraded simultaneously. IMO there shouldn't exist a scenario in which we upgrade the compute node and sink writers separately. Maybe the motivation for `SinkWriterV1Adapter` is to help us port existing sink writers to the new trait gradually?

**Comment by wenym1** [comment]
> Maybe the motivation for SinkWriterV1Adapter is to help us port existing sink writers to the new trait gradually

Exactly. `backward compatibility` is what I meant for it, and it may not be that accurate.

**Review by StrikeW — APPROVED** [review]
LGTM

**Pull request merged**